docs/madr/decisions/039-meshservice-api.md

```diff
@@ -0,0 +1,119 @@
+# MeshService API
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/6331
+
+## Context and Problem Statement
+
+Currently, our definition of "service" is derived from `kuma.io/service` tag placed in a Dataplane object. It has a several problems.
+
+**Scalability**
+* There are a number of objects whose size grows `O(#services)`: ServiceInsights, VIP ConfigMap, ZoneIngress availableServices. At scale storing objects of this size becomes a problem. There are limits for a size of an object in Kube API server.
+* Listing available services requires iterating through all Dataplane objects.
+* Separating Envoy endpoint generation from cluster generation. Right now a single change of Dataplane object would trigger cluster generation even though the cluster may not have changed.
+
+**Indirection**
+* There is no way to add metadata to the service. You can add metadata to a workload (Dataplane).
+  While we can compute shared metadata of workloads of same `kuma.io/service` and consider this a service metadata, there is a confusing indirection.
+
+## Considered Options
+
+* Shard objects
+* MeshService
+
+## Decision Outcome
+
+Chosen option: "MeshService", because it helps us to provide better foundation to build things on top of it. 
+
+### Positive Consequences
+
+* Solved scalability problems
+
+### Negative Consequences
+
+* A pretty big effort (implementation, a strategy of migration).
+* Temporarily increased complexity of a product until the migration is done.
+
+## Pros and Cons of the Options
+
+### Shard objects
+
+We could shard objects that are `O(#services)`, although it's not clear by what key.
+While this solves the problem of scalability, it does not solve other "side" problems that can be solved with a dedicated MeshService object.
+
+### MeshService
+
+We introduce a MeshService object.
+
+```yaml
+kind: MeshService
+metadata:
+  name: redis
+  namespace: redis-system
+  labels: # you could then select this in when you use `MeshService` in `to` section
+    team: db-operators
+    kuma.io/mesh: default
+spec:
+  selector:
+    dataplaneTags: # tags in Dataplane object, see below
+      app: redis
+      k8s.kuma.io/namespace: redis-system # added automatically
+      kuma.io/zone: east-1 # added automatically
+  ports:
+  - port: 6739
+    targetPort:
+      value: 6739
+    appProtocol: tcp
+  - name: some-port
+    port: 16739
+    targetPort:
+      name: target-port-from-container # name of the inbound
+    appProtocol: tcp
```

Comment:
k8s has both `appProtocol` and `protocol`, are we fine having only `appProtocol`?

Reply:
yes, we care only about appProtocol